pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,476 @@
+	// remote ctx here and trace it.
+	if tv, created := r.tryStore(r.newTraceValue(te, nil, nil)); created {
+		tv.logf(1, "registering remote trace %s", tv)
+	}
+}
+
+// MaybeRegister is called on an entry that has been proposed to raft. This will
+// begin logging all subsequent updates to this entry. It returns true if the
+// registration is successful. A duplicate registration of the same index is
+// considered a success and returns true, however the older registration is kept
+// and this registration is ignored.
+func (r *RaftTracer) MaybeRegister(ctx context.Context, ent raftpb.Entry) bool {
+	// If the index is nil, then we can't trace this entry. This can happen if
+	// there is a leader/leaseholder spilt. We don't have an easy way to handle
+	// this today, so don't attempt to trace it.
+	if ent.Index == 0 {
+		log.VEvent(ctx, 2, "skip registering raft proposal without index")
+		return false
+	}
+
+	// Only register the entry if this is a traced context with verbose logging.
+	span := tracing.SpanFromContext(ctx)
+	if span == nil || span.RecordingType() != tracingpb.RecordingVerbose {
+		return false
+	}
+
+	// This must be the last conditional. If this returns true we must call
+	// storeEntryWithTracing to not leak a registered permit.
+	if !r.reserveSpace() {
+		log.VEvent(ctx, 2, "too many active raft traces, skipping")
+		return false
+	}
+
+	ctx, span = r.tracer.StartSpanCtx(ctx, "raft trace",
+		tracing.WithParent(span), tracing.WithFollowsFrom())
+	if tv, created := r.tryStore(
+		r.newTraceValue(kvserverpb.TracedEntry{
+			Index:   kvpb.RaftIndex(ent.Index),
+			TraceID: span.TraceID(),
+			SpanID:  span.SpanID(),
+		},
+			ctx,
+			span,
+		)); created {
+		tv.logf(1, "registering local trace %s", tv)
+	}
+	return true
+}
+
+// MaybeTrace logs the message in every trace it is relevant to.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []kvserverpb.TracedEntry {
+	// NB: This check is an optimization to handle the common case where there
+	// are no registered traces on this replica.
+	if r.numRegisteredReplica.Load() == 0 {
+		return nil
+	}
+
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// removeEntry removes the trace at the given index and decrements the
+// registered counters at the replica and store level.
+func (r *RaftTracer) removeEntry(index kvpb.RaftIndex) {
+	tv, found := r.m.LoadAndDelete(index)
+	if !found {
+		return
+	}
+	// Don't allow additional tracing to this context.
+	r.destroy(tv)
+}
+
+func (r *RaftTracer) destroy(tv *traceValue) {
+	tv.mu.Lock()
+	defer tv.mu.Unlock()
+
+	if tv.mu.propSpan != nil {
+		tv.mu.propSpan.Finish()
+		tv.mu.propCtx = nil
+		tv.mu.propSpan = nil
+	}
+	r.numRegisteredReplica.Add(-1)
+	r.numRegisteredStore.Add(-1)
```

Comment:
Could probably move these decrements to the front, so that we hold the mutex slightly shorter. The counters can be decremented asap to allow new registrations immediately.

Reply:
Good point. I moved them first before the mu lock.